src/DynamoCore/DSEngine/EngineController.cs

```diff
@@ -51,9 +48,53 @@ private void OnTraceReconciliationComplete(TraceReconciliationEventArgs e)
         private readonly Queue<GraphSyncData> graphSyncDataQueue = new Queue<GraphSyncData>();
         private readonly Queue<List<Guid>> previewGraphQueue = new Queue<List<Guid>>();
         public bool VerboseLogging;
+
         private readonly Object macroMutex = new Object();
 
-        public static CompilationServices CompilationServices; 
+        public static CompilationServices CompilationServices;
+
+        /// <summary>
+        /// Get DesignScript core.
+        /// </summary>
+        public ProtoCore.Core LiveRunnerCore
+        {
+            get
+            {
+                return liveRunnerServices.Core;
+            }
+        }
+
+        /// <summary>
+        /// Get DesignScript runtime core.
+        /// </summary>
+        public ProtoCore.RuntimeCore LiveRunnerRuntimeCore
+        {
+            get
+            {
+                return liveRunnerServices.RuntimeCore;
+            }
+        }
+
+
+        /// <summary>
+        /// Return libary service instance.
+        /// </summary>
+        public LibraryServices LibraryServices
+        {
+            get { return libraryServices; }
+        }
+
+        public CodeCompletionServices CodeCompletionServices
+        {
+            get { return codeCompletionServices; }
+        }
+
+        /// <summary>
+        /// A property defining whether the EngineController has been disposed or not.
+        /// This is a conservative field, as there should only be one owner of a valid
```

Comment:
In an ideal world, only HomeWorkspaceModel would have an instance of this class.  However, it's passed around willy-nilly and I have the feeling it might be disposed without telling everyone.  One fix would be to add a Disposed event, but that just adds more state and more disposal issues.  This is set as soon as we enter the Dispose code block.